plutus-playground-lib/src/Playground/Schema.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE AllowAmbiguousTypes   #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RecordWildCards       #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# OPTIONS_GHC -fno-warn-unticked-promoted-constructors #-}
+
+{-| This module handles exposing a Contract API to the Plutus Playground frontend.
+
+In practice this means having a way of turning the Contract's effect
+rows into a Schema declaration, by using an 'EndpointToSchema'
+instance.
+
+For each row, we either expose it as a schema (if it's an endpoint),
+or ignore it (if it's a known effect that doesn't have an equivalent
+frontend call.
+|-}
+module Playground.Schema
+    ( endpointsToSchemas
+    , EndpointToSchema
+    ) where
+
+import           Data.Row                                        (Empty, KnownSymbol, Label (Label))
+import           Data.Row.Internal                               (LT ((:->)), Row (R))
+import qualified Data.Text                                       as Text
+import           Language.Plutus.Contract.Effects.ExposeEndpoint (ActiveEndpoints, EndpointValue)
+import           Language.Plutus.Contract.Schema                 ()
+import           Playground.Types                                (EndpointName (EndpointName), FunctionSchema (FunctionSchema, argumentSchema, functionName))
+import           Schema                                          (FormSchema, ToSchema, toSchema)
+
+class EndpointToSchema (s :: Row *) where
+    endpointsToSchemas :: [FunctionSchema FormSchema]
+
+instance EndpointToSchema Empty where
+    endpointsToSchemas = []
+
+instance (ToSchema params, KnownSymbol l, EndpointToSchema (R bs)) =>
+         EndpointToSchema (R ((l :-> (EndpointValue params, ActiveEndpoints)) : bs)) where
+    endpointsToSchemas = FunctionSchema {..} : endpointsToSchemas @(R bs)
+      where
+        functionName = EndpointName . Text.pack . show $ Label @l
+        argumentSchema = [toSchema @params]
+
+instance (EndpointToSchema (R bs)) =>
+         EndpointToSchema (R ("address" :-> response : bs)) where
```

Comment:
Instead of filtering these out here in `EndpointToSchema s`, you could also call `EndpointToSchema (s .\\ BlockchainActions)`. There's an example of this in `Language.Plutus.Contract.App` l. 75

Also, the symbols have type synonyms defined which would be a bit nicer: for example `Language.Plutus.Contract.Effects.WatchAddress.AddressSymbol`

Reply:
Much nicer, thanks.